### ISSUE TITLE
Long-term user information cache to improve performance of authentication requests

### ISSUE THREAD

**Issue by fredreichbier** [privacyidea/privacyidea#670]
In a productive setting, it is important to optimize the performance of authentication requests. Processing authentication requests also entails user information lookup. As user information are unlikely to change very often, we could improve performance by adding an long-term user information cache, i.e. caching information for days or weeks. Some thoughts:

* The LDAP resolver already implements a simple, but limited cache: It is implemented as a global Python dictionary, which limits its scope to the current process and may incur problems with race conditions.
* We could use privacyIDEA's SQL database to implement the cache. This way, we would also benefit from database replication.
* We could implement the cache independently from the utilized ID resolver.
* We should decide which user information should be part of the cache.
* We should identify situations in which the cache should be invalidated, e.g. changing resolver settings, realm settings, ...
* We should provide a way to invalidate the cache manually (e.g. from a cronjob).

**Comment by cornelinux** [comment]
# Which data to cache

The token lookup needs a **User-Object**
* userid
* resolver
* realm

The **authentication** is performed with:
* loginname
* realm (optinoal)
* resolver (optional)

Thus the minimum data to **cache** in the SQL table is:
* loginname
* resolver
* realm
* userid

For other operations additional user attributes **could** be cached. But these are not relevant for authentication.

# Cache invalidation

The following actions need to invalidate the cache:

* save or delete a resolver
* save or delete a realm
* create a user
* delete a user

**Comment by cornelinux** [comment]
In ``lib.users.py`` we probably need to decorate the following:

* get_username
* Users.get_resolvers
* Users.get_user_identifiers

**Comment by fredreichbier** [comment]
e71e1afa0a771ddaa772eb63493188b43e6d4297 implements the decorators:

* get_username performs a cache lookup, but does not populate the cache. This is because the function does not know the realm which should be added to the cache.
* User.get_resolvers and User.get_user_identifiers both perform cache lookup and also populate the cache

I guess cache invalidation might be the hardest task :-) Some open questions:

* When should expired entries be removed from the database? We could remove an expired entry when we encounter it (i.e. when we try to read from the cache and only find an expired entry), or we rely on the user to periodically prune expired entries.
* How do we implement cache invalidation? I tend to a decorator-based solution on the lib level: We could just decorate the relevant functions below ``privacyidea.lib`` (e.g. ``realm.set_realm``) with a decorator ``@invalidate_cache`` which invalidates the cache after (or before?) entering the function.
* What should happen if we invalidate the cache? Should we invalidate only relevant entries or prune the whole ``userinfo_cache`` table? I tend to the latter, because things could get really messy if we fail to invalidate some relevant entries.

**Comment by cornelinux** [comment]
* The query in ``one_or_none`` method in lib/usercache.py will always return one result or none. Why not use
  the sqlalchemy function ``query().filter().first()`` instead of ``.all()?``

**Comment by cornelinux** [comment]
@fredreichbier watch out: I renamed the branch ``670/user-information-cache`` to ``670/user-cache``!

**Comment by cornelinux** [comment]
# Some thoughts again on what to cache

If the administrator rearranges the realm. Delete a realm, rename a realm, the cache users are not valid anymore. Although the information of (userid, username, resolver) is still valid.

We could only cache this tuple (userid, username, resolver). To avoid querying the resolvers...?

# Deleting and expiring

## Deleting

Cache user data should be deleted when 
* Renaming: No action necessary. "Renaming" the resolver creates a new resolver.
* Deleting the resolver. When a resolver is deleted, we have the name of the resolver. Then we should delete all entries of this resolver from the cache.
* Saving a resolver. When a resolver is saved, we do not know, if the configuration of the resolver has changed. Thus we should delete all entries of this resolver from the cache.

We could have a function ``delete_user_cache(resolver=resolvername)``.

## Expiring

When saving user to the cache table, we could check for expired entries. But this would be an additional DB call on each request. This might be time consuming...

We could simply delete from the database table all entries with a to old timestamp. This could be one single DELETE statement. We probably should not call it during a validate/check? This might be time consuming? When should we call it?

We could have a function ``delete_user_cache(expired=True)``.

**Comment by fredreichbier** [comment]
The branch now also implements cache invalidation. Some things I've noticed:

* Currently we invalidate the cache entries belonging to a specific user when we delete it. Should we also do that when we edit a user?
* Some tests use something like `cache_username(get_username, uid, resolver)`. But this will actually invoke `cache_username` twice, as `get_username` is already decorated! Maybe we need to come up with a better solution for that.
* I'm not really happy with the "inlined" cache invalidation in 6db8a097be955e4e6925d6fc7a49d7f8ddfe5384, but I though using a decorator (as I initially had in mind) would make things a bit confusing. What do you think?

I'll work on improving and extending the unit tests for the user cache feature tomorrow.

**Comment by cornelinux** [comment]
* Edit User

   Yes, we probably should delete the user from the cache, when the user is edited. Since the *loginname* could also be an editable attribute.

* cache_username:

  Does it call cache_username twice? We need to pass *any* function as first parameter to cache_username. I would not spent to much time on this test. In the end it tests the code nicely.

* inline delete_user_cache
  
   I think this is totally fine.

**Comment by cornelinux** [comment]
Last change looks good to me.
Test also looks quite nice.

As I see we have the invalidation of the cache on:
* resolver save
* resolver delete
* user edit/delete

So I guess we are fine?

**Comment by cornelinux** [comment]
# Documentation

we should add some documentation here, where the cache timeout is configured:
https://github.com/privacyidea/privacyidea/blob/670/user-cache/doc/configuration/system_config.rst

like 

>  here you can configure the timeout for the resolver cache. If set to =0, the resolver cache is deactivated.

and a more in depth explanation here:
https://github.com/privacyidea/privacyidea/blob/670/user-cache/doc/configuration/useridresolvers.rst

like what is happening with the cache. 
* Note side effects with several resolvers in the cache.
* Note: expired entries will not be deleting. => when enlonging the timeout, you will see old entires again.
* Note: Use an external cron job to clean up the cache, i.e. delete expired entries.

**Comment by fredreichbier** [comment]
I'm not 100% happy with aaa23a9215a96f57703e24e537d6d7402534212d, as we now have `_get_resolvers` and `user_init`, which do pretty much the same. But I guess it's okay for now.

I think the cache invalidation points should be fine! We should probably also make sure that the functions to which we added invalidation are not bypassed anywhere (e.g. if `IdResolver.update_user` is called directly, the cache entries are not invalidated).

I'll write some documentation the next days! I also think it would be nice to be able to quantify the speedup we gained (not just for this issue, but generally), but I'm not really sure about how to do that yet (maybe I'll open another issue for this).
